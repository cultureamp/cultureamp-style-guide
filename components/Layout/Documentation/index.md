---
imports:
  Demo: components/Demo.js
  IntroParagraph: components/IntroParagraph.js
  presets: ./_presets.js
---

# Layout

<IntroParagraph>

A standard Culture Amp layout with left-hand-side navigation bar, sidebar, and optional headers / footers.

Currently available for React.

</IntroParagraph>

<Demo presets={presets} />
