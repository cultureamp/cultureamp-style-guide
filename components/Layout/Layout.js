// @flow
import * as React from 'react';
import styles from './Layout.module.scss';

type LayoutProps = {
  children: React.Node,
};

class Layout extends React.Component<LayoutProps> {
  render() {
    let content = [];
    let navbar = null;
    let header = null;
    let sidebar = null;
    let footer = null;

    React.Children.toArray(this.props.children).forEach(child => {
      if (!child) return;
      if (child.type.name == NavigationBar.name) {
        navbar = child;
      } else if (child.type.name == Header.name) {
        header = child;
      } else if (child.type.name == Sidebar.name) {
        sidebar = child;
      } else if (child.type.name == Footer.name) {
        footer = child;
      } else {
        content.push(child);
      }
    });

    return (
      <div className={styles.root}>
        {navbar}
        <div className={styles.page}>
          {header}
          <div className={styles.body}>
            {sidebar}
            <div className={styles.content}>{content}</div>
          </div>
          {footer}
        </div>
      </div>
    );
  }

  static NavigationBar = NavigationBar;
  static Sidebar = Sidebar;
  static Header = Header;
  static Footer = Footer;
}

function NavigationBar(props: { children: React.Node }) {
  return <div className={styles.navigationBar}>{props.children}</div>;
}

function Sidebar(props: { children: React.Node }) {
  return <div className={styles.sidebar}>{props.children}</div>;
}

function Header(props: { children: React.Node }) {
  return <div className={styles.header}>{props.children}</div>;
}

function Footer(props: { children: React.Node }) {
  return <div className={styles.footer}>{props.children}</div>;
}

export default Layout;
