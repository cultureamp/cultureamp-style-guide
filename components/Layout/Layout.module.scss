@import './styles';

.root {
  @extend %root;
}

.navigationBar {
  @extend %navigation-bar;
  box-sizing: border-box;
}

.page {
  @extend %page;
  box-sizing: border-box;
  position: relative;
}

.header {
  @extend %header;
  box-sizing: border-box;
}

.body {
  @extend %body;
  box-sizing: border-box;
}

.bodyInner {
  @extend %body-inner;
}

.sidebar {
  @extend %sidebar;
  box-sizing: border-box;
}

.content {
  @extend %content;
  box-sizing: border-box;
}

.footer {
  @extend %footer;
  box-sizing: border-box;
}

.toasts {
  position: absolute;
  top: 0;
  right: 0;
}

.announcers {
  position: absolute;
  top: 0;
  left: 0;
  opacity: 0.5;
}
