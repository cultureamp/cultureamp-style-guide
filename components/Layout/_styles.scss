@import '~cultureamp-style-guide/styles/layers';
@import '~cultureamp-style-guide/styles/color';
@import '~cultureamp-style-guide/styles/type';

/*
      +-------------------------------------+
      |n|         header                    |
      |a|-----------------------------------| --
      |v|       |                           |  |
      |b|       |                           |  |
      |a|   s   |                           |  |
      |r|   i   |                           |  |
      | |   d   |                           |  |
      | |   e   |         content           |  |  body
      | |   b   |                           |  |
      | |   a   |                           |  |
      | |   r   |                           |  |
      | |       |                           |  |
      | |       |                           |  |
      | |-----------------------------------| --
      | |         footer                    |
      +-------------------------------------+

        |-----------------------------------|
                      page
*/

%root {
  display: flex;
  align-items: stretch;
}

%navigation-bar {
  flex: 0 0 auto;
}

%page {
  display: flex;
  flex: 1 0 0px;
  flex-direction: column;
  height: 100vh; // scrolling content will overflow
}

%header,
%footer {
  flex: 0 0 auto;
}

%body {
  // %body provides a positioning context for %body-inner,
  // which may be nested n levels deep in the DOM.
  position: relative;
  flex: 1 1 0px;
  height: 0; // flex-basis 0px doesn't work in most browsers

  // to support %legacy-sticky-footer (TODO: remove once it's not used)
  display: flex;
  flex-direction: column;
}

// old Murmur pages include this as a child of %body
// NOTE: not compatible with %body-inner - cannot use both on same page
%legacy-sticky-footer {
  margin-top: auto;
  width: 100%;
}

// nested (possibly deeply) descendant of %body that will adopt its dimensions
// and support a sticky footer that conforms to those dimensions
%body-inner {
  display: flex;

  // match position & dimensions of %body
  position: absolute; // required for height to work when content is short
  height: 100%;
  width: 100%;

  @media print {
    display: block;
    position: static;
    height: auto;
    width: auto;
  }
}

$sidebar-width: 12 * $ca-grid;
%sidebar {
  background-color: $ca-palette-stone;
  height: 100vh;
  overflow-x: hidden;
  overflow-y: auto;
  position: fixed;
  width: $sidebar-width;
  z-index: $ca-z-index-fixed;
  padding: $ca-grid;

  // stack sidebar and content vertically in print
  @media print {
    background-color: transparent;
    overflow-y: visible;
    overflow-x: visible;
    position: static;
    width: auto;
    height: auto;
    padding: 0;
  }
}

// enforce fixed width to avoid change when scrollbar on %sidebar is visible
%sidebar-inner {
  width: $sidebar-width - 2 * $ca-grid;
}

%content {
  // to support sticky footer
  display: flex;
  flex-direction: column;
  margin-left: $sidebar-width;
  padding: $ca-grid;

  flex: 1 1 auto;

  // stack sidebar and content vertically in print
  @media print {
    display: block;
    padding: 0;
    margin-left: 0;
    overflow: visible;
  }

  &:only-child {
    margin: 0 auto;
    width: 100%; // fixes centering in IE 11 at max-width
    min-width: 100%; // prevents too-wide content slipping under NavigationBarApp

    @media (min-width: 480px) {
      min-width: auto;
      max-width: 900px;
    }
  }
}
