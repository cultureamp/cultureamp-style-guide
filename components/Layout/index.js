// @flow
export { default } from './Layout.js';
