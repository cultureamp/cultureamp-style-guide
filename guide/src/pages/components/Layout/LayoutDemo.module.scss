@import '~cultureamp-style-guide/styles/color';
@import '~cultureamp-style-guide/styles/type';
@import '~cultureamp-style-guide/styles/layers';
@import '~cultureamp-style-guide/styles/border';

$padding: $ca-grid;

.layoutWrapper {
  box-sizing: border-box;
  padding: 0 $padding;
  width: 100%;
  height: 100vh;
  overflow: scroll;
  position: relative;

  > div {
    background: $ca-palette-stone;
  }

  //Get the header and sidebar to not be "position: fixed" for the purpose of this demo.
  header {
    position: absolute;
  }
}

.sidebar {
  h2 {
    @include ca-type-ideal-display;
  }
  p {
    @include ca-type-ideal-body;
  }
}

.content {
  h1 {
    @include ca-type-ideal-page-title;
  }
  p {
    @include ca-type-ideal-lede;
  }
}

.flash {
  @include ca-type-ideal-body;
  @include ca-inherit-baseline;
  text-align: center;
  position: absolute;
  top: $ca-grid;
  right: $ca-grid;
  padding: $ca-grid;
  width: 300px;
  border: $ca-border-color 1px solid;
  border-radius: $ca-border-radius;
  box-shadow: $ca-box-shadow;
  background-color: $ca-palette-stone;
  z-index: $ca-z-index-popover;
}

.footer {
  @include ca-type-ideal-small;
  text-align: center;
}
