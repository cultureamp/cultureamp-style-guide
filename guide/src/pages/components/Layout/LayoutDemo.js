import React from 'react';
import { Layout } from '@cultureamp/kaizen/components/GlobalIALayout';
import styles from './LayoutDemo.module.scss';

// Note: because of the `position: fixed` styling, we need to wrap the NavigationBar and insert some styles to make it position: relative.
const LayoutDemo = props => (
  <div className={styles.layoutWrapper}>
    <Layout {...props} />
  </div>
);

export default LayoutDemo;
