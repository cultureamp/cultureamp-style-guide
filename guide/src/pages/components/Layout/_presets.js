// @flow
import React from 'react';
import Layout from 'cultureamp-style-guide/components/Layout';
import NavigationBar from 'cultureamp-style-guide/components/NavigationBar';
import Icon from 'cultureamp-style-guide/components/Icon';
import homeIcon from 'cultureamp-style-guide/icons/home.svg';
import caIcon from 'cultureamp-style-guide/icons/ca-monogram.svg';
import supportIcon from 'cultureamp-style-guide/icons/support.svg';
import LayoutDemo from './LayoutDemo';
import styles from './LayoutDemo.module.scss';

const navBar = (
  <Layout.NavigationBar>
    <NavigationBar>
      <NavigationBar.Link href="/" text="Home" active={true} />
      <NavigationBar.Link
        href="http://academy.cultureamp.com/"
        text="Support"
      />
      <NavigationBar.Menu
        items={[
          {
            label: 'About Culture Amp',
            link: 'https://www.cultureamp.com/',
          },
          {
            label: 'Contribute to this guide',
            link:
              'https://github.com/cultureamp/cultureamp-style-guide/tree/master/guide',
            target: '_blank',
          },
        ]}
      >
        <div
          style={{
            backgroundColor: '#ffffff',
            color: '#F8A6AE',
            borderRadius: '3px',
          }}
        >
          <Icon
            icon={caIcon}
            role="img"
            title="Culture Amp Logo"
            inheritSize={true}
          />
        </div>
      </NavigationBar.Menu>
    </NavigationBar>
  </Layout.NavigationBar>
);

const sidebar = (
  <Layout.Sidebar>
    <div className={styles.sidebar}>
      <h2>Actions</h2>
      <p>Go here...</p>
      <h2>Filters</h2>
      <p>Go here...</p>
      <p>
        Note: Usually the navigation bar and sidebar will not scroll as the main
        page content scrolls, but we have disabled the fixed positioning for
        this demo.
      </p>
    </div>
  </Layout.Sidebar>
);

const content = (
  <div className={styles.content}>
    <h1>This is the main content</h1>
    <p>It will scroll with the main window</p>
    <p>...</p>
    <p>...</p>
    <p>...</p>
    <p>...</p>
    <p>...</p>
    <p>...</p>
    <p>...</p>
    <p>...</p>
    <p>...</p>
    <p>...</p>
    <p>...</p>
    <p>...</p>
    <p>...</p>
    <p>...</p>
    <p>...</p>
    <p>While the sidebar and navbar should remain fixed.</p>
  </div>
);

const header = (
  <Layout.Header>
    <div className={styles.flash}>Welcome back!</div>
  </Layout.Header>
);

const footer = (
  <Layout.Footer>
    <div className={styles.footer}>&copy; 2018 Culture Amp</div>
  </Layout.Footer>
);

const presets = [
  {
    name: 'With sidebar',
    node: (
      <LayoutDemo>
        {navBar}
        {sidebar}
        {content}
      </LayoutDemo>
    ),
  },
  {
    name: 'Without sidebar',
    node: (
      <LayoutDemo>
        {navBar}
        {content}
      </LayoutDemo>
    ),
  },
  {
    name: 'With footer and header',
    node: (
      <LayoutDemo>
        {header}
        {navBar}
        {sidebar}
        {content}
        {footer}
      </LayoutDemo>
    ),
  },
];

export default presets;
