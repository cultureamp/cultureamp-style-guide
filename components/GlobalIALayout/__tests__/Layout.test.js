import React from 'react';
import { mount } from 'enzyme';
import Layout from '../Layout';

test('The layout components have the correct displayNames', () => {
  expect(Layout.displayName).toBe('Layout');
  expect(Layout.NavigationBar.displayName).toBe('NavigationBar');
  expect(Layout.Sidebar.displayName).toBe('Sidebar');
  expect(Layout.Header.displayName).toBe('Header');
  expect(Layout.Footer.displayName).toBe('Footer');
  expect(Layout.Toasts.displayName).toBe('Toasts');
  expect(Layout.Announcers.displayName).toBe('Announcers');
});

test('The NavigationBar component renders correctly', () => {
  const component = mount(<Layout.NavigationBar />);
  expect(component).toMatchSnapshot();
});

test('The Sidebar component renders correctly', () => {
  const component = mount(<Layout.Sidebar />);
  expect(component).toMatchSnapshot();
});

test('The Header component renders correctly', () => {
  const component = mount(<Layout.Header />);
  expect(component).toMatchSnapshot();
});

test('The Footer component renders correctly', () => {
  const component = mount(<Layout.Footer />);
  expect(component).toMatchSnapshot();
});

test('The Toasts component renders correctly', () => {
  const component = mount(<Layout.Toasts />);
  expect(component).toMatchSnapshot();
});

test('The Announcers component renders correctly', () => {
  const component = mount(<Layout.Announcers />);
  expect(component).toMatchSnapshot();
});

test('The Layout component renders correctly', () => {
  const component = mount(<Layout />);
  expect(component).toMatchSnapshot();
});

test('The Layout component renders child components in the correct positions', () => {
  const component = mount(
    <Layout>
      <Layout.NavigationBar>NavigationBar</Layout.NavigationBar>
      <Layout.Sidebar>Sidebar</Layout.Sidebar>
      <Layout.Header>Header</Layout.Header>
      <Layout.Footer>Footer</Layout.Footer>
      <Layout.Toasts>Toasts</Layout.Toasts>
      <Layout.Announcers>Announcers</Layout.Announcers>
      <div>My content</div>
    </Layout>
  );
  expect(component).toMatchSnapshot();
});
