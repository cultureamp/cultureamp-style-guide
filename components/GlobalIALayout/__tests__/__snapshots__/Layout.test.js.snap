// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Announcers component renders correctly 1`] = `
<Announcers>
  <div
    aria-live="assertive"
    className="announcers"
  />
</Announcers>
`;

exports[`The Footer component renders correctly 1`] = `
<Footer>
  <footer
    className="footer"
  />
</Footer>
`;

exports[`The Header component renders correctly 1`] = `
<Header>
  <aside
    className="header"
  />
</Header>
`;

exports[`The Layout component renders child components in the correct positions 1`] = `
<Layout>
  <div
    className="root"
  >
    <NavigationBar
      key=".0"
    >
      <div
        className="navigationBar"
      >
        NavigationBar
      </div>
    </NavigationBar>
    <Announcers
      key=".5"
    >
      <div
        aria-live="assertive"
        className="announcers"
      >
        Announcers
      </div>
    </Announcers>
    <div
      className="page"
    >
      <Header
        key=".2"
      >
        <aside
          className="header"
        >
          Header
        </aside>
      </Header>
      <Toasts
        key=".4"
      >
        <div
          aria-live="assertive"
          className="toasts"
        >
          Toasts
        </div>
      </Toasts>
      <div
        className="body"
      >
        <div
          className="bodyInner"
        >
          <Sidebar
            key=".1"
          >
            <div
              className="sidebar"
            >
              <div
                className="sidebarInner"
              >
                Sidebar
              </div>
            </div>
          </Sidebar>
          <main
            className="content"
          >
            <div
              key=".6"
            >
              My content
            </div>
          </main>
        </div>
      </div>
      <Footer
        key=".3"
      >
        <footer
          className="footer"
        >
          Footer
        </footer>
      </Footer>
    </div>
  </div>
</Layout>
`;

exports[`The Layout component renders correctly 1`] = `
<Layout>
  <div
    className="root"
  >
    <div
      className="page"
    >
      <div
        className="body"
      >
        <div
          className="bodyInner"
        >
          <main
            className="content"
          />
        </div>
      </div>
    </div>
  </div>
</Layout>
`;

exports[`The NavigationBar component renders correctly 1`] = `
<NavigationBar>
  <div
    className="navigationBar"
  />
</NavigationBar>
`;

exports[`The Sidebar component renders correctly 1`] = `
<Sidebar>
  <div
    className="sidebar"
  >
    <div
      className="sidebarInner"
    />
  </div>
</Sidebar>
`;

exports[`The Toasts component renders correctly 1`] = `
<Toasts>
  <div
    aria-live="assertive"
    className="toasts"
  />
</Toasts>
`;
